### PR TITLE
boot: Add support for stack protector/trapping math/ubsan

### PR DESCRIPTION
**Pull request by medhefgo**

### CODE REVIEW

**Comment by medhefgo** [comment]
This should hopefully make us a little bit safer.

(Also, not really sure about the merits of the last commit. I actually dislike the repetition those warnings cause.)

**Comment by bluca** [comment]
With the recent changes, is the NX bit dealt with? It's going to become a requirement and firmwares are going to enforce it this year

**Comment by medhefgo** [comment]
> With the recent changes, is the NX bit dealt with? It's going to become a requirement and firmwares are going to enforce it this year

It should be, unless our compilers are lying to us.

Though, I believe this is just about NX enforcement on the memory mappings in general and not about the NX support bit indicated with the PE characteristic. Because I cannot see anything in the EDK2 sources to that effect: the memory page allocation for PE image only requests code pages, so any data sections still reside in executable pages.

**Comment by medhefgo** [comment]
I managed to add ubsan support.

**Comment by DaanDeMeyer** [comment]
@medhefgo Can you rebase past the test-time-util failures?

**Comment by bluca** [comment]
> > With the recent changes, is the NX bit dealt with? It's going to become a requirement and firmwares are going to enforce it this year
> 
> It should be, unless our compilers are lying to us.
> 
> Though, I believe this is just about NX enforcement on the memory mappings in general and not about the NX support bit indicated with the PE characteristic. Because I cannot see anything in the EDK2 sources to that effect: the memory page allocation for PE image only requests code pages, so any data sections still reside in executable pages.

The full requirements are detailed here: https://techcommunity.microsoft.com/t5/hardware-dev-center/updated-uefi-signing-requirements/ba-p/1062916#:~:text=The%20following%20list%20contains%20the%20latest%20requirements%20for,certificate%20and%20an%20Azure%20Active%20Directory%20%28AAD%29%20account.

> Effective 11/30/2022 all submissions must satisfy and attest to NX compatibility as described by: [New UEFI CA memory mitigation requirements for signing](https://techcommunity.microsoft.com/t5/hardware-dev-center/new-uefi-ca-memory-mitigation-requirements-for-signing/ba-p/3608714)
> - Section Alignment of the submitted PE file must be aligned with 4KB page size. 
> - Section Flags must not combine IMAGE_SCN_MEM_WRITE and IMAGE_SCN_MEM_EXECUTE for any given section.
> - DLL Characteristics must include IMAGE_DLLCHARACTERISTICS_NX_COMPAT

**Comment by medhefgo** [comment]
We *should* be in the clear regarding those requirements. I would sure love to test this, though.

**Review by bluca — APPROVED** [review]

File: src/boot/efi/log.c [hunk 1]
```diff
@@ -81,6 +81,12 @@ void __stack_chk_fail_local(void) {
         __stack_chk_fail();
 }
 
+/* Called by libgcc for some fatal errors like integer overflow with -ftrapv. */
+_used_ _noreturn_ void abort(void);
+void abort(void) {
```
Comment (yuwata) [inline src/boot/efi/log.c:86]: Why `_used_` is sometimes set to prototype but sometimes to function?

Comment (medhefgo) [reply src/boot/efi/log.c:86]: To annoy people?

**Review by yuwata — APPROVED** [review]

**Pull request merged**